File: project1/minesweeper.c
/************************************************************************
 * minesweeper.c														*
 *																		*
 * Author(s): Jaxon Wright, Chad Teitsma  								*
 ***********************************************************************/

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define BOARD_SIZE_MIN 5
#define BOARD_SIZE_MAX 15
#define PCT_MINES_MIN 10
#define PCT_MINES_MAX 70

#define RCN_COLOR	"\x1b[38;2;180;180;180m"
#define MIN_COLOR	"\x1b[31m"
#define ZER_COLOR   "\x1b[36m"
#define ONE_COLOR	"\x1b[38;2;156;122;255m"
#define TWO_COLOR 	"\x1b[38;2;95;171;97m"
#define THR_COLOR	"\x1b[38;2;255;118;109m"
#define FOU_COLOR 	"\x1b[38;2;115;105;189m"
#define FIV_COLOR	"\x1b[38;2;184;93;90m"
#define SIX_COLOR	"\x1b[38;2;67;144;146m"
#define SEV_COLOR	"\x1b[38;2;191;238;124m"
#define EIG_COLOR	"\x1b[38;2;255;171;251m"
#define CLE_COLOR	"\x1b[39;49m"

typedef struct {
	bool is_mine;
	int mines;
	bool visible;
} Cell;

typedef enum {WON, LOST, INPROGRESS} Status;

/************************************************************************
 * YOU MUST NOT DEFINE ANY GLOBAL VARIABLES (i.e., OUTSIDE FUNCTIONS).  *
 * COMMUNICATION BETWEEN FUNCTIONS MUST HAPPEN ONLY VIA PARAMETERS.     *
 ************************************************************************/

/************************************************************************
 * Function declarations/prototypes										*
 ************************************************************************/
void printColored(Cell);

void displayMenu();

int getBoardSize();

int getPercentMines();

void initBoard(int size, Cell board[][size]);

void placeMinesOnBoard(int size, Cell board[][size], int nbrMines);

void fillInMineCountForNonMineCells(int size, Cell board[][size]);

int nbrOfMines(int size, Cell board[][size]);

int getNbrNeighborMines(int row, int col, int size, Cell board[][size]);

void displayBoard(int size, Cell board[][size], bool displayMines);

Status selectCell(int row, int col, int size, Cell board[][size]);

int nbrVisibleCells(int size, Cell board[][size]);

//void setImmediateNeighborCellsVisible(int row, int col, int size, Cell board[][size]);

void setAllNeighborCellsVisible(int row, int col, int size, Cell board[][size]);


/************************************************************************
 * Main driver of the program. Uses the functions defined above.		*
 ************************************************************************/
int main()
{
	int row, col, size, nbrMines;
	char command;
	bool displayMines = false;
	Status gameState = INPROGRESS;

	printf("************************************************************************************************************************\n");
	printf("* _|      _|  _|_|_|  _|      _|  _|_|_|_|    _|_|_|  _|          _|  _|_|_|_|  _|_|_|_|  _|_|_|    _|_|_|_|  _|_|_|   *\n");
	printf("* _|_|  _|_|    _|    _|_|    _|  _|        _|        _|          _|  _|        _|        _|    _|  _|        _|    _| *\n");
	printf("* _|  _|  _|    _|    _|  _|  _|  _|_|_|      _|_|    _|    _|    _|  _|_|_|    _|_|_|    _|_|_|    _|_|_|    _|_|_|   *\n");
	printf("* _|      _|    _|    _|    _|_|  _|              _|    _|  _|  _|    _|        _|        _|        _|        _|    _| *\n");
	printf("* _|      _|  _|_|_|  _|      _|  _|_|_|_|  _|_|_|        _|  _|      _|_|_|_|  _|_|_|_|  _|        _|_|_|_|  _|    _| *\n");
	printf("* a game by jaxon wright and chad teitsma                                                                              *\n");
	printf("************************************************************************************************************************\n\n");
	size = getBoardSize();

	// declare 2D array of cells
	Cell board[size][size];

	initBoard(size,board);

	// determine number of mine to place on the board
	nbrMines = (int) (size * size * (getPercentMines()/100.0));

	// place mines randomly on the board
	placeMinesOnBoard(size,board,nbrMines);

	// For each non-mine cell, sets the neighboring mine count
	fillInMineCountForNonMineCells(size,board);

	displayBoard(size,board,displayMines);

	while (true) {
		printf("Enter command (m/M for command menu): ");
		scanf("%c",&command);
		getchar();

		switch (command) {
			case 'm': case 'M':
				displayMenu();
				break;

			case 'c': case 'C':
				do {
					printf("Enter row and column of cell: ");
					scanf("%d%d",&row,&col);
					getchar();
					if (row < 1 || row > size || col < 1 || col > size) {
						printf("Invalid row or column values. Try again.\n");
					}
				} while (row < 1 || row > size || col < 1 || col > size);
				row--;
				col--;
				gameState = selectCell(row,col,size,board);
				displayBoard(size,board,displayMines);
				break;

			case 's': case 'S':
				displayMines = true;
				displayBoard(size,board,displayMines);
				break;

			case 'h': case 'H':
				displayMines = false;
				displayBoard(size,board,displayMines);
				break;

			case 'b': case 'B':
				displayBoard(size,board,displayMines);
				break;

			case 'q': case 'Q':
				printf("Bye.\n");
				return 0;

			default:
				printf("Invalid command. Try again.\n");
		}

		// display appropriate message if the game is over
		if (gameState == WON) {
			printf("You found all the mines. Congratulations. Bye.");
			return 0;
		} else if (gameState == LOST) {
			printf("Oops. Sorry, you landed on a mine. Bye");
			return 0;
		}
	}

	return 0;
}

/************************************************************************
 * Displays list of commands allowed at the prompt						*
 ************************************************************************/
void displayMenu()
{
	printf("List of available commands:\n");
	printf("   Show Mines: s/S\n");
	printf("   Hide Mines: h/H\n");
	printf("   Select Cell: c/C\n");
	printf("   Display Board: b/B\n");
	printf("   Display Menu: m/M\n");
	printf("   Quit: q/Q\n\n");
}

/************************************************************************
 * Initializes the fields of each cell on the board as follows:			*
 * 		is_mine field to false											*
 * 		mines field to 0												*
 * 		visible field to false											*
 ************************************************************************/
void initBoard(int size, Cell board[][size])
{
	for (int i = 0; i < size; i++){
		for (int j = 0; j < size; j++){
			Cell cell;
			
			cell.is_mine = false;
			cell.mines = 0;
			cell.visible = false;
			
			board[i][j] = cell;
		}
	}
}

/************************************************************************
 * Places the specified number of mines randomly on the board			*
 ************************************************************************/
void placeMinesOnBoard(int size, Cell board[][size], int nbrMines){
	srand(time(NULL));
	int x, y;
	int count = 0;
	
	while (count < nbrMines){
		x = rand()%(size);
		y = rand()%(size);
		
		//if there is not a mine already at the randomly generated spot then 
		//put one there
		if(!board[x][y].is_mine){
			board[x][y].is_mine = true;
			count++;
		}
	}
}

/************************************************************************
 * For each non-mine cell on the board, set the "mines" field to the	*
 * number of mines in the immediate neighborhood.						*
 ************************************************************************/
void fillInMineCountForNonMineCells(int size, Cell board[][size]) {
	for (int i = 0; i < size; i++){
		for (int j = 0; j < size; j++){
			if(!board[i][j].is_mine)
				board[i][j].mines = getNbrNeighborMines(i,j,size,board);
		}
	}
}

/************************************************************************
 * Counts and returns the number of mines on the board					*
 ************************************************************************/
int nbrOfMines(int size, Cell board[][size]) {
	int count = 0;
	
	for (int i = 0; i < size; i++){
		for (int j = 0; j < size; j++){
			if (board[i][j].is_mine)
				count++;
		}
	}

	return count;
}

/************************************************************************
 * Returns the number of mines in the immediate neighborhood of a cell	*
 * at location (row,col) on the board.									*
 ************************************************************************/
int getNbrNeighborMines(int row, int col, int size, Cell board[][size]) {
	int count = 0;
	
	//check left
	if (col > 0 && board[row][col-1].is_mine){
		count++;
	}
	//check diag up left
	if ((col > 0 && row > 0) && board[row-1][col-1].is_mine){
		count++;
	}
	//check above
	if (row > 0 && board[row-1][col].is_mine){
		count++;
	}
	//check diag up right
	if ((row > 0 && col < size-1) && board[row-1][col+1].is_mine){
		count++;
	}
	//check right
	if (col < size-1 && board[row][col+1].is_mine){
		count++;
	}
	//check diag down right
	if ((row < size-1 && col < size-1) && board[row+1][col+1].is_mine){
		count++;
	}
	//check below
	if (row < size-1 && board[row+1][col].is_mine){
		count++;
	}
	//check diag down left
	if ((row < size-1 && col > 0) && board[row+1][col-1].is_mine){
		count++;
	}
	
	return count;
}

/************************************************************************
 * Displays the board. If a cell is not currently visible and has a		*
 * mine, show the mine if the displayMines is true. Used for debugging	*
 * and testing purposes.												*
 ************************************************************************/
void displayBoard(int size, Cell board[][size], bool displayMines) {	
	//Print top row, displaying column numbers
	printf("\t");
	for (int i=1; i <= size; i++)
		printf(RCN_COLOR "%d\t" CLE_COLOR,i);
	
	printf("\n");
	//print row numbers and rest of board
	for (int i=0; i < size; i++){
		//Stuff is getting colored lulz
		printf(RCN_COLOR "%d\t" CLE_COLOR,i+1);
		for (int j=0; j < size; j++){
			if(board[i][j].is_mine && displayMines)
				printf(MIN_COLOR "*\t" CLE_COLOR);
			else if (board[i][j].visible)
				printColored(board[i][j]);
			else
				printf("?\t");
		}	
		printf("\n");
	}
}

void printColored(Cell board){
	switch(board.mines){
		case 0:
			printf(ZER_COLOR "%d\t" CLE_COLOR, 0);
			break;
		case 1:
			printf(ONE_COLOR "%d\t" CLE_COLOR, 1);
			break;
		case 2:
			printf(TWO_COLOR "%d\t" CLE_COLOR, 2);
			break;
		case 3:
			printf(THR_COLOR "%d\t" CLE_COLOR, 3);
			break;
		case 4:
			printf(FOU_COLOR "%d\t" CLE_COLOR, 4);
			break;
		case 5:
			printf(FIV_COLOR "%d\t" CLE_COLOR, 5);
			break;
		case 6:
			printf(SIX_COLOR "%d\t" CLE_COLOR, 6);
			break;
		case 7:
			printf(SEV_COLOR "%d\t" CLE_COLOR, 7);
			break;
		case 8:
			printf(EIG_COLOR "%d\t" CLE_COLOR, 8);
			break;
	}
}

/************************************************************************
 * Prompts the user for board size, reads and validates the input		*
 * entered, and returns the integer if it is within valid range.		*
 * repeats this in a loop until the user enters a valid value.			*
 ************************************************************************/
int getBoardSize() {
	int size = 0;

	printf("Enter the board size ( %d to %d ): ", BOARD_SIZE_MIN, BOARD_SIZE_MAX); 
	
	//scans from user input to character array
	char board_size[50];
  	fgets(board_size,50,stdin);
  	//convert from char array to integer. If it is not a number returns a negative int
  	size = atoi(board_size);

	if (size < BOARD_SIZE_MIN || size > BOARD_SIZE_MAX){
		printf("Invalid input.\n");
		size = getBoardSize();
	}
	return size;
}

/************************************************************************
 * Prompts the user for percentage of mines to place on the board,		*
 * reads and validates the input entered, and returns the integer if it	*
 * is within valid range. repeats this in a loop until the user enters	*
 * a valid value for board size.										*
 ************************************************************************/
int getPercentMines() {
	int percent = 0;

	printf("Enter the percentage of mines on the board ( %d to %d ): ", PCT_MINES_MIN, PCT_MINES_MAX); 
	
	//scans from user input to character array
	char board_size[50];
  	fgets(board_size,50,stdin);
  	//convert from char array to integer. If it is not a number returns a negative int
  	percent = atoi(board_size);

	if (percent < PCT_MINES_MIN|| percent > PCT_MINES_MAX){
		printf("Invalid input.\n");
		percent = getPercentMines();
	}
	return percent;
}

/************************************************************************
 * Process cell selection by user during the game						*
 ************************************************************************/
Status selectCell(int row, int col, int size, Cell board[][size]){
	
	board[row][col].visible = true;
	
	if (board[row][col].is_mine){
		return LOST;
	}
	if (board[row][col].mines == 0) {
		setAllNeighborCellsVisible(row, col, size, board);
	}
	if(nbrOfMines(size, board) + nbrVisibleCells(size, board) == size*size) {
		return WON;
	}

	return INPROGRESS;
}

/************************************************************************
 * Returns the number of cells that are currently visible.				*
 ************************************************************************/
int nbrVisibleCells(int size, Cell board[][size]){
	int count = 0;

	for (int i = 0; i < size; i++){
		for (int j = 0; j < size; j++){
			if (board[i][j].visible)
				count++;
		}
	}
	
	return count;
}

/************************************************************************
 * If the mine count of a cell at location (row,col) is zero, then make	*
 * the cells in the immediate neighborhood visible and repeat this		*
 * process for each of the cells in this set of cells that have a mine	*
 * count of zero, and so on.											*
 ************************************************************************/
void setAllNeighborCellsVisible(int row, int col, int size, Cell board[][size]){
	
	//check left
	if (col > 0 && !board[row][col-1].visible)
		selectCell(row, col-1, size, board);
	//check diag up left
	if ((col > 0 && row > 0) && !board[row-1][col-1].visible)
		selectCell(row-1, col-1, size, board);
	//check above
	if (row > 0 && !board[row-1][col].visible)
		selectCell(row-1, col, size, board);
	//check diag up right
	if ((row > 0 && col < size-1) && !board[row-1][col+1].visible)
		selectCell(row-1, col+1, size, board);
	//check right
	if (col < size-1 && !board[row][col+1].visible)
		selectCell(row, col+1, size, board);
	//check diag down right
	if ((row < size-1 && col < size-1) && !board[row+1][col+1].visible)
		selectCell(row+1, col+1, size, board);
	//check below
	if (row < size-1 && !board[row+1][col].visible)
		selectCell(row+1, col, size, board);
	//check diag down left
	if ((row < size-1 && col > 0) && !board[row+1][col-1].visible)
		selectCell(row+1, col-1, size, board);
	
}

File: project6/minesweeper.py
#
#  Minesweeper game
#
#  Author(s): Jaxon Wright and Chad Teitsma
#


# Constants
WON = 0
LOST = 1
INPROGRESS = 2
BOARD_SIZE_MIN = 5
BOARD_SIZE_MAX = 15
PCT_MINES_MIN = 10
PCT_MINES_MAX = 70
RCN_COLOR = "\033[38;5;245m"
MIN_COLOR = "\033[38;5;160m"
ZER_COLOR = "\033[38;5;248m"
ONE_COLOR = "\033[38;5;98m"
TWO_COLOR = "\033[38;5;70m"
THR_COLOR = "\033[38;5;209m"
FOU_COLOR = "\033[38;5;111m"
FIV_COLOR = "\033[38;5;209m"
SIX_COLOR = "\033[38;5;153m"
SEV_COLOR = "\033[38;5;115m"
EIG_COLOR = "\033[38;5;134m"
CLE_COLOR = "\x1b[39;49m"

class Cell:
    def __init__(self,nbr_mines,is_mine,visible):
        self.nbr_mines = nbr_mines
        self.is_mine = is_mine
        self.visible = visible
    
class Minesweeper:
    
    # initialize a minesweeper object
    def __init__(self,board_size, percent_mines):
        self.board_size = board_size
        self.nbr_mines = int(board_size*board_size*(percent_mines/100.0))
        
         # setup a 2-dimensional array of Cell objects
        self.board = [[0 for x in range(self.board_size)] for x in range(self.board_size)]
        for i in range(self.board_size):
            for j in range(self.board_size):
                self.board[i][j]=Cell(0,False,False)
                
        self.place_mines_on_board()
        self.fill_in_minecount_for_non_mine_cells()
        
    # places mines randomly on the board   
    def place_mines_on_board(self):
        count = 0
        while count < self.nbr_mines:
            x = random.randrange(self.board_size)
            y = random.randrange(self.board_size)
            
            if not self.board[x][y].is_mine:
                self.board[x][y].is_mine = True
                count += 1
        
    # for each non-mine cell on the board, set @nbr_mines of each Cell on the board
    # to the number of mines in the immediate neighborhood.
    def fill_in_minecount_for_non_mine_cells(self):
        for i in range(self.board_size):
            for j in range(self.board_size):
                if not self.board[i][j].is_mine:
                    self.board[i][j].nbr_mines = self.get_nbr_neighbor_mines(i,j)
    
    # processes cell selection by user during the game
    # returns WON, LOST, or INPROGRESS
    def select_cell(self,row,col):
        self.board[row][col].visible = True
        
        if self.board[row][col].is_mine:
            return LOST
            
        if self.board[row][col].nbr_mines == 0:
            self.set_all_neighbor_cells_visible(row,col)
            
        if self.nbr_mines + self.nbr_visible_cells() == self.board_size*self.board_size:
            return WON
            
        return INPROGRESS
    
    # returns the number of mines in the immediate neighborhood of a cell
    # at location (row,col) on the board.
    def get_nbr_neighbor_mines(self,row,col):
        count = 0
    
        #check left
      	if col > 0 and self.board[row][col-1].is_mine:
      		count += 1
      	#check diag up left
      	if (col > 0 and row > 0) and self.board[row-1][col-1].is_mine:
      		count += 1
      	#check above
      	if (row > 0 and self.board[row-1][col].is_mine):
      		count += 1
      	#check diag up right
      	if (row > 0 and col < self.board_size-1) and self.board[row-1][col+1].is_mine:
      		count += 1
      	#check right
      	if col < self.board_size-1 and self.board[row][col+1].is_mine:
      		count += 1
      	#check diag down right
      	if (row < self.board_size-1 and col < self.board_size-1) and self.board[row+1][col+1].is_mine:
      		count += 1
      	#check below
      	if row < self.board_size-1 and self.board[row+1][col].is_mine:
      		count += 1
      	#check diag down left
      	if (row < self.board_size-1 and col > 0) and self.board[row+1][col-1].is_mine:
      		count += 1
     
      	return count
    
    # returns the number of cells that are currently visible on the board
    def nbr_visible_cells(self):
        count = 0
        
        for i in range(self.board_size):
            for j in range(self.board_size):
                if self.board[i][j].visible:
                    count += 1
        
        return count
        
    # if the mine count of a cell at location (row,col) is zero, then make 
    # the cells in the immediate neighborhood visible and repeat this
    # process for each of the cells in this set of cells that have a mine
    # count of zero, and so on.
    def set_all_neighbor_cells_visible(self,row,col):
        #check left
        if col > 0 and not self.board[row][col-1].visible:
        	self.select_cell(row, col-1)
        #check diag up left
        if (col > 0 and row > 0) and self.board[row-1][col-1].visible:
        	self.select_cell(row-1, col-1)
        #check above
        if (row > 0 and not self.board[row-1][col].visible):
        	self.select_cell(row-1, col)
        #check diag up right
        if ((row > 0 and col < self.board_size-1) and not self.board[row-1][col+1].visible):
        	self.select_cell(row-1, col+1)
        #check right
        if (col < self.board_size-1 and not self.board[row][col+1].visible):
        	self.select_cell(row, col+1)
        #check diag down right
        if ((row < self.board_size-1 and col < self.board_size-1) and not self.board[row+1][col+1].visible):
        	self.select_cell(row+1, col+1)
        #check below
        if (row < self.board_size-1 and not self.board[row+1][col].visible):
        	self.select_cell(row+1, col)
        #check diag down left
        if ((row < self.board_size-1 and col > 0) and not self.board[row+1][col-1].visible):
        	self.select_cell(row+1, col-1)
    
    # returns a string representation of the board
    def to_string(self,display_mines):
        result = ""
        
        #top row
        for i in range(self.board_size):
            if i == 0:
                result += RCN_COLOR
                result += "%6d" % (i+1)
            else:
                result += "%3d" % (i+1)
        result += "\n"
            
        #the rest    
        for i in range(self.board_size):
            #row label
            result += RCN_COLOR
            result += "%3d" % (i+1)
            result += CLE_COLOR
            for j in range(self.board_size):
                if self.board[i][j].visible:
                    if self.board[i][j].is_mine:
                        result += MIN_COLOR + "  *" + CLE_COLOR
                    else:
                        result += self.print_colored(self.board[i][j].nbr_mines)
                else:
                    if display_mines and self.board[i][j].is_mine:
                        result += MIN_COLOR + "  *" + CLE_COLOR
                    else:
                        result += "  ?"
            result += "\n"
            
        return result
    
    def print_colored(self,nbr_mines):
        if nbr_mines == 0:
            return ZER_COLOR + "  0" + CLE_COLOR
        if nbr_mines == 1:
            return ONE_COLOR + "  1" + CLE_COLOR
        if nbr_mines == 2:
            return TWO_COLOR + "  2" + CLE_COLOR
        if nbr_mines == 3:
            return THR_COLOR + "  3" + CLE_COLOR
        if nbr_mines == 4:
            return FOU_COLOR + "  4" + CLE_COLOR
        if nbr_mines == 5:
            return FIV_COLOR + "  5" + CLE_COLOR
        if nbr_mines == 6:
            return SIX_COLOR + "  6" + CLE_COLOR
        if nbr_mines == 7:
            return SEV_COLOR + "  7" + CLE_COLOR
        if nbr_mines == 8:
            return EIG_COLOR + "  8" + CLE_COLOR

def get_board_size():
    board_size = 0
    while board_size < BOARD_SIZE_MIN or board_size > BOARD_SIZE_MAX:
        board_size = raw_input("Enter board size ("+str(BOARD_SIZE_MIN)+" .. "+str(BOARD_SIZE_MAX)+"): ")
        try:
            board_size = int(board_size)
            if  board_size < BOARD_SIZE_MIN or board_size > BOARD_SIZE_MAX:
                print "Invalid board size. Number out of range. Try again."
        except ValueError:
            print("Invalid board size. Must be a number. Try again.")

    return board_size

def get_percent_mines():
    percent_mines = 0
    while percent_mines < PCT_MINES_MIN or percent_mines >PCT_MINES_MAX:
        percent_mines = raw_input("Enter percentage of mines on the board ("+str(PCT_MINES_MIN)+" .. "+str(PCT_MINES_MAX)+"): ")
        try:
            percent_mines = int(percent_mines)
            if  percent_mines < PCT_MINES_MIN or percent_mines > PCT_MINES_MAX:
                print "Invalid value for percentage. Number out of range. Try again."
        except ValueError:
            print("Invalid value for percentage. Must be a number. Try again.")

    return percent_mines
    
def display_menu():
    print "List of available commands:"
    print "   Show Mines: s/S"
    print "   Hide Mines: h/H"
    print "   Select Cell: c/C"
    print "   Display Board: b/B"
    print "   Display Menu: m/M"
    print "   Quit: q/Q"


def main():
    
    print ONE_COLOR
    print "************************************************************************************************************************"
    print "* _|      _|  _|_|_|  _|      _|  _|_|_|_|    _|_|_|  _|          _|  _|_|_|_|  _|_|_|_|  _|_|_|    _|_|_|_|  _|_|_|   *"
    print "* _|_|  _|_|    _|    _|_|    _|  _|        _|        _|          _|  _|        _|        _|    _|  _|        _|    _| *"
    print "* _|  _|  _|    _|    _|  _|  _|  _|_|_|      _|_|    _|    _|    _|  _|_|_|    _|_|_|    _|_|_|    _|_|_|    _|_|_|   *"
    print "* _|      _|    _|    _|    _|_|  _|              _|    _|  _|  _|    _|        _|        _|        _|        _|    _| *"
    print "* _|      _|  _|_|_|  _|      _|  _|_|_|_|  _|_|_|        _|  _|      _|_|_|_|  _|_|_|_|  _|        _|_|_|_|  _|    _| *"
    print "* a game by jaxon wright and chad teitsma....this time in python!                                                      *"
    print "************************************************************************************************************************\n"
    print CLE_COLOR
    
    display_mines = False
    game_state = INPROGRESS
    
    
    board_size = get_board_size()
    percent_mines = get_percent_mines()
    board = Minesweeper(board_size,percent_mines)
    
    print board.to_string(display_mines)
    
    while True:
        command = raw_input("Enter command (m/M for menu): ")
        str(command)
        
        if command == 'm' or command == 'M':
            display_menu()
            
        elif command == 'c' or command == 'C':
            row = 0
            col = 0
            while row < 1 or row > board_size or col < 1 or col > board_size:
                line = raw_input("Enter row and column of cell: ")
                coordinate = line.split()
                input_error = False
                try:
                    row = coordinate[0]
                    col = coordinate[1]
                    row = int(row)
                    col = int(col)
                except ValueError:
                    print "Invalid row or column values. Must be a number. Try again."
                    input_error = True
                except IndexError:
                    print "Invalid row or column values. Please enter 2 values. Try again."
                    input_error = True
                
                if  (row < 1 or row > board_size or col < 1 or col > board_size) and not input_error:
                    print "Invalid row or column values. Out of bounds. Try again."
                    
            game_state = board.select_cell(row-1,col-1)
            print board.to_string(display_mines)
            
        elif command == 's' or command == 'S':
             display_mines = True
             print board.to_string(display_mines)
             
        elif command == 'h' or command == 'H':
            display_mines = False
            print board.to_string(display_mines)
        
        elif command == 'b' or command == 'B':
            print board.to_string(display_mines)
            
        elif command == 'q' or command == 'Q':
            print "Bye."
            return
        
        else:
            print "Invalid command. Try again."
        #end of elif's    
            
        if game_state == WON:
            print  "You found all the mines. Congratulations. Bye."
            return
        elif game_state == LOST:
            print "Oops. Sorry, you landed on a mine. Bye"
            return
    #end of while
    
    #end of main
                    
# run the game
main()
    
        